### PR TITLE
Add internal user authz

### PR DESCRIPTION
**Pull request by petersbingham**
**What this PR does / why we need it**:
``` Detailed description of PR; If no description, remove the block.

```

**Special notes for your reviewer**:
``` If no notes, remove the block.

```

**Release note**:
``` Release notes; If no release note is required, just write "NONE" within the block.

```

## Summary by CodeRabbit

# Release Notes

* **New Features**
  * Introduced role-based authorization framework for internal operations, including dedicated roles for task processing, tenant provisioning, and workflow management.

* **Bug Fixes**
  * Fixed authorization bypass where system operations previously bypassed access control checks.

* **Chores**
  * Refactored authorization enforcement across async tasks, managers, and API controllers to use the new role-based system.

### CODE REVIEW

File: internal/authz/repo_internal_policies.go [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+	Policies []BasePolicy[constants.InternalRole, RepoResourceTypeName, RepoAction]
+}
+
+var InternalRepoPolicyData = internalRepoPolicies{
+	Roles: []constants.InternalRole{
+		constants.InternalTenantProvisioningRole,
+	},
+	Policies: []BasePolicy[constants.InternalRole, RepoResourceTypeName, RepoAction]{
+		NewPolicy(
+			"InternalTaskProcessing",
```
Comment (petersbingham) [inline internal/authz/repo_internal_policies.go]: Update name here from cpy/paste.

File: internal/authz/repo_base.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+// all actions which are used in policies which can be performed on resource types
+const (
+	RepoActionList   RepoAction = "list"
+	RepoActionFirst  RepoAction = "first"
+	RepoActionCount  RepoAction = "count"
+	RepoActionCreate RepoAction = "create"
+	RepoActionUpdate RepoAction = "update"
+	RepoActionDelete RepoAction = "delete"
+
+	// This is a special action, allowing the internal user to circumnavigate
```
Comment (petersbingham) [inline internal/authz/repo_base.go:45]: I don't like the idea of "dynamic". I am thinking of just replacing these with role checks (in this PR). This is what the workflows are using. See CheckInternalUserRole in check.go.
The preference would be to have all the permission specified in the policies but there is no neat way of achieving this with the current abstraction, I think we can deal with this better later.

File: utils/context/context.go [hunk 3]
```diff
@@ -168,24 +171,64 @@ func ExtractClientDataAuthContext(ctx context.Context) (map[string]string, error
...
+	if !ok || internalRole == "" {
+		return "", ErrExtractClientData
 	}
 
-	clientDataCopy := *clientData
-	clientDataCopy.Identifier = uuid.Max.String()
+	return internalRole, nil
+}
+
+func ExtractSource(ctx context.Context) (string, error) {
```
Comment (minh-nghia) [inline utils/context/context.go:227]: Np: Maybe `ExtractClientDataSource`? Source alone is quite broad.

File: internal/middleware/clientdata.go
Comment (minh-nghia) [inline internal/middleware/clientdata.go:104]: `validateGroupRoles` should be using `InjectInternalClientData` to be sure.

**Review by minh-nghia — COMMENTED** [review]
Will also need to cover event reconciler, mainly these functions:
resolveTasks, confirmJob, jobDoneFunc, jobFailedFunc, jobCanceledFunc

**Comment by anettlippert** [comment]
@coderabbitai /summarize

File: cmd/tenant-manager-cli/commands/commands.go [hunk 4]
```diff
@@ -44,13 +44,13 @@ func NewCommandFactory(
 	}
 
 	cm := manager.NewCertificateManager(ctx, r, svcRegistry, cfg)
-	um := manager.NewUserManager(r, cmkAuditor)
+	um := manager.NewUserManager(r, nil, cmkAuditor)
```
Comment (anettlippert) [inline cmd/tenant-manager-cli/commands/commands.go:47]: the AuthzLoader is quite often nil, not only here but also at other calls, maybe we should introduce two functions: NewUserManagerWithoutAuthz and NewUserManager? Or at least comment why it isn't needed

File: cmd/task-worker/main.go [hunk 5]
```diff
@@ -122,7 +122,7 @@ func registerTasks(
 
 	r := sql.NewRepository(dbCon)
 
-	sis, err := manager.NewSystemInformationManager(r, svcRegistry, &cfg.ContextModels.System)
+	sis, err := manager.NewSystemInformationManager(r, nil, svcRegistry, &cfg.ContextModels.System)
```
Comment (anettlippert) [inline cmd/task-worker/main.go:133]: same here, maybe we can have two functions: NewSystemInformationManager NewSystemInformationManagerWithoutAuthz so it is more clear

File: internal/authz/repo_base.go (hunk 2 above)
Comment (anettlippert) [reply internal/authz/repo_base.go:45]: yes, let us do this later

File: internal/auditor/auditor.go [hunk 6]
```diff
@@ -63,8 +64,8 @@ func (a *Auditor) getEventMetadata(ctx context.Context) (otlpaudit.EventMetadata
 
 	userInitID, err := cmkcontext.ExtractClientDataIdentifier(ctx)
 	if err != nil {
-		// If no user context is available, use default max UUID
-		userInitID = constants.SystemUser.String()
+		// If no user context is available, use empty string
```
Comment (anettlippert) [inline internal/auditor/auditor.go:67]: userInitID isn't an empty string, it is max UUID as placeholder, so maybe you can change the comment

File: internal/authz/authorization_data.go [hunk 7]
```diff
@@ -7,64 +7,109 @@ import (
 	"github.com/openkcm/cmk/internal/errs"
 )
 
-type AuthorizationKey[TResourceTypeName, TAction comparable] struct {
-	TenantID         TenantID
-	UserGroup        string
+type BusinessUserAuthzKey struct {
```
Comment (anettlippert) [inline internal/authz/authorization_data.go:10]: I don't like the therm user here in the name of that struct, as it doesn't contain any user info. Same for InternalUserAuthzKey can we maybe use it without User?

**Review by anettlippert — CHANGES_REQUESTED** [review]

File: utils/context/context.go (hunk 3 above)
Comment (petersbingham) [reply utils/context/context.go:227]: `ClientData` is specific to business users. I will change to `ExtractUserSource`. Not perfect but ideal change would likely require changing the clientdata name, which I don't want to touch.

Comment (petersbingham) [reply utils/context/context.go:227]: I've reworked all the naming. Now `BusinessUser` and `InternalUser` throughout, makes things much clearer I think. `ClientData` still referenced from the common sdk and in the configuraiton but these' too hard t change now.

File: internal/middleware/clientdata.go
Comment (petersbingham) [reply internal/middleware/clientdata.go:104]: This is being set in the `GetRoleFromIAM` function.

File: cmd/tenant-manager-cli/commands/commands.go (hunk 4 above)
Comment (petersbingham) [reply cmd/tenant-manager-cli/commands/commands.go:47]: This made sense but no lnger relevant, as I've removed the dependency to address another comment.

File: cmd/task-worker/main.go (hunk 5 above)
Comment (petersbingham) [reply cmd/task-worker/main.go:133]: This also applies to the `NewSystemManager`. I think the problem is more that this really should have the loader and not `nil`, so there is basically missing authz here, which I've fixed. Thanks for pointing this out.

There are still instances in the _test_ code where this is `nil`. Based on previous comment by @jmpTeixeira02 the preference is to avoid lists of specifically named functions (although I'll leave Joao to comment if the context is different here) and also 1) now that this only affects test code I don't think worth introducing another function for this; 2) other parameters are set to `nil` in the test code (eg `eventsFactory` for `newSystemManager`).
